packages/beacon-node/src/api/impl/validator/index.ts

```diff
@@ -648,50 +661,73 @@ export function getValidatorApi(
+            return assembleBuilderBlockResponse(blockAttributes, currentState, commonBlockBody, builderBlockBody);
+          })
+          .finally(() => {
+            if (builderTimer) builderTimer({source: ProducedBlockSource.builder});
+          })
       : Promise.reject(new Error("Builder disabled"));
 
-    const enginePromise = isEngineEnabled
-      ? produceEngineFullBlockOrContents(slot, randaoReveal, graffiti, {
-          feeRecipient,
-          strictFeeRecipientCheck,
-          // skip checking and recomputing head in these individual produce calls
-          skipHeadChecksAndUpdate: true,
-          commonBlockBody,
-          parentBlockRoot,
-        }).then((engineBlock) => {
-          // Once the engine returns a block, in the event of either:
-          // - suspected builder censorship
-          // - builder boost factor set to 0 or builder selection `executionalways`
-          // we don't need to wait for builder block as engine block will always be selected
-          if (
-            engineBlock.shouldOverrideBuilder ||
-            builderBoostFactor === BigInt(0) ||
-            builderSelection === routes.validator.BuilderSelection.ExecutionAlways
-          ) {
-            controller.abort();
-          }
-          return engineBlock;
-        })
+    const engineBodyPromise = isEngineEnabled
+      ? produceEngineBlockBody({...blockAttributes, currentState}, loggerContext)
+          .then(async (engineBlockBody) => {
+            const commonBlockBody = await commonBlockBodyPromiseFn();
+            const resp = await assembleEngineBlockResponse(
+              blockAttributes,
+              currentState,
+              commonBlockBody,
+              engineBlockBody,
+              {
+                strictFeeRecipientCheck,
+                feeRecipient,
+              }
+            );
+
+            // Once the engine returns a block, in the event of either:
+            // - suspected builder censorship
+            // - builder boost factor set to 0 or builder selection `execution-always`
+            // we don't need to wait for builder block as engine block will always be selected
+            if (
+              engineBlockBody.shouldOverrideBuilder ||
+              builderBoostFactor === BigInt(0) ||
+              builderSelection === routes.validator.BuilderSelection.ExecutionAlways
+            ) {
+              controller.abort();
+            }
+            return resp;
+          })
+          .finally(() => {
+            if (engineTimer) engineTimer({source: ProducedBlockSource.engine});
+          })
       : Promise.reject(new Error("Engine disabled"));
 
-    const [builder, engine] = await resolveOrRacePromises([builderPromise, enginePromise], {
-      resolveTimeoutMs: cutoffMs,
-      raceTimeoutMs: BLOCK_PRODUCTION_RACE_TIMEOUT_MS,
-      signal: controller.signal,
-    });
+    const [_, [builder, engine]] = await Promise.all([
+      commonBlockBodyPromiseFn(),
```

Comment:
I don't see why we have to await for `commonBlockBodyPromiseFn()` here
if both `builder` and `engine` rejects, we throw error in the below code
if one of them resolve, the `commonBlockBodyPromiseFn()` function should have been fulfilled

removing it would also keep the old code unchanged

Reply:
+1